src/test/java/online/bookstore/book/BookControllerTest.java

```diff
@@ -0,0 +1,221 @@
+                objectMapper.readTree(jsonResponse).get("content").toString(),
+                new TypeReference<List<BookDto>>() {}
+        );
+
+        for (int i = 0; i < books.size(); i++) {
+            BookDto expectedBook = books.get(i);
+            BookDto actualBook = actualBooks.get(i);
+            EqualsBuilder.reflectionEquals(expectedBook, actualBook, "categories");
+        }
+    }
+
+    @Test
+    @WithMockUser(username = "user", roles = {"USER"})
+    void getById_WithValidId_ShouldReturnBookDto() throws Exception {
+        Long bookId = 1L;
+        BookDto expected = new BookDto()
+                .setId(bookId)
+                .setTitle("Book One")
+                .setAuthor("Author One")
+                .setPrice(BigDecimal.valueOf(19.99))
+                .setIsbn("978-3-16-148410-0")
+                .setDescription("Description for book one")
+                .setCoverImage("http://example.com/book1.jpg")
+                .setCategories(Set.of());
+        when(bookService.getById(bookId)).thenReturn(expected);
+        MvcResult result = mockMvc.perform(get("/books/{id}", bookId)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andReturn();
+        String jsonResponse = result.getResponse().getContentAsString();
+        BookDto actual = objectMapper.readValue(jsonResponse, BookDto.class);
+        EqualsBuilder.reflectionEquals(expected, actual, "categories");
+    }
+
+    @Test
+    @WithMockUser(username = "admin", roles = {"ADMIN"})
+    @Sql(
+            scripts = "classpath:database/books/delete-new-book.sql",
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD
+    )
+    @Sql(
+            scripts = "classpath:database/books/delete-fiction-category.sql",
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD
+    )
+    @Transactional
+    void createBook() throws Exception {
+        CreateBookRequestDto requestBook = new CreateBookRequestDto();
+        requestBook.setTitle("New Book");
+        requestBook.setAuthor("Author");
+        requestBook.setPrice(BigDecimal.TEN);
+        requestBook.setIsbn("978-3-16-148410-0");
+        requestBook.setDescription("Description");
+        requestBook.setCoverImage("http://example.com/cover3.jpg");
+        Category category = new Category();
+        category.setId(1L);
+        category.setName("Fiction");
+        category.setDescription("Fictional books");
+        requestBook.setCategories(Set.of(category));
+
+        BookDto expected = new BookDto();
+        expected.setTitle(requestBook.getTitle());
+        expected.setAuthor(requestBook.getAuthor());
+        expected.setPrice(requestBook.getPrice());
+        expected.setIsbn(requestBook.getIsbn());
+        expected.setDescription(requestBook.getDescription());
+        expected.setCoverImage(requestBook.getCoverImage());
+        expected.setCategories(requestBook.getCategories());
```

Comment:
better move such logic to the private method or util class and simplify your test cases.